firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,285 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
     pub fn flush_dirty(&self) -> Option<()> {
         self.store.flush_dirty()
     }
+
+    pub fn get_iter<K: AsRef<[u8]>>(
+        &self,
+        key: Option<K>,
+        root: DiskAddress,
+    ) -> Result<MerkleKeyValueStream<'_, S>, MerkleError> {
+        // TODO: if DiskAddress::is_null() ...
+        Ok(MerkleKeyValueStream {
+            key_state: IteratorState::new(key),
+            merkle_root: root,
+            merkle: self,
+        })
+    }
+}
+
+enum IteratorState<'a> {
+    /// Start iterating at the beginning of the trie,
+    /// returning the lowest key/value pair first
+    StartAtBeginning,
+    /// Start iterating at the specified key
+    StartAtKey(Vec<u8>),
+    /// Continue iterating after the given last_node and parents
+    Iterating {
+        last_node: ObjRef<'a>,
+        parents: Vec<(ObjRef<'a>, u8)>,
+    },
+}
+impl IteratorState<'_> {
+    fn new<K: AsRef<[u8]>>(starting: Option<K>) -> Self {
+        match starting {
+            None => Self::StartAtBeginning,
+            Some(key) => Self::StartAtKey(key.as_ref().to_vec()),
+        }
+    }
+}
+
+// The default state is to start at the beginning
+impl<'a> Default for IteratorState<'a> {
+    fn default() -> Self {
+        Self::StartAtBeginning
+    }
+}
+pub struct MerkleKeyValueStream<'a, S> {
+    key_state: IteratorState<'a>,
+    merkle_root: DiskAddress,
+    merkle: &'a Merkle<S>,
```

Comment:
🤔 this could be problematic. In the RPC API, expects that we create and cache iterators for RPC clients... If the iterator has an immutable borrow of `Merkle`, then we can't write to the trie while the iterator exists. 

We should leave a note about this in the doc-comments.

Reply:
Will the merkle tree stay immutable even after traversed one time? I think it should, as we will use a new iterator for a changed merkle?